Assets/Scripts/Switch.cs

```diff
@@ -84,34 +88,41 @@ private bool IsSwitchActive
     void Start()
     {
         animator = GetComponent<Animator>();
-        animator.SetInteger(Animator.StringToHash(animColorBool),
-            (int)switchColor);
-        puzzle = connectedPuzzle.GetComponent<IPuzzle>();
-        puzzle.AppendSwitchList(this);
+        animator.SetInteger(animParamColor, (int)switchColor);
         audioSource = GetComponent<AudioSource>();
     }
 
+    private void OnEnable()
+    {
+        Puzzle.SolutionIsIncorrect += ResetByPuzzle;
+    }
+
+    private void OnDisable()
+    {
+        Puzzle.SolutionIsIncorrect -= ResetByPuzzle;
```

Comment:
Make an event handler method called OnEVENTNAME. Then make a method called "ResetPuzzle" that you call from the event handler.